src/main/java/no/unit/nva/amazon/s3/CreateUploadHandler.java

```diff
@@ -0,0 +1,151 @@
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static no.unit.nva.amazon.s3.GatewayResponse.ACCESS_CONTROL_ALLOW_ORIGIN;
+import static no.unit.nva.amazon.s3.GatewayResponse.BODY_KEY;
+import static org.apache.http.HttpHeaders.CONTENT_TYPE;
+import static org.apache.http.HttpStatus.SC_BAD_REQUEST;
+import static org.apache.http.HttpStatus.SC_CREATED;
+import static org.apache.http.HttpStatus.SC_INTERNAL_SERVER_ERROR;
+import static org.apache.http.entity.ContentType.APPLICATION_JSON;
+
+public class CreateUploadHandler implements RequestHandler<Map<String, Object>, GatewayResponse> {
+
+
+    public static final String ALLOWED_ORIGIN_KEY = "ALLOWED_ORIGIN";
+    public static final String AWS_REGION_KEY = "AWS_REGION";
+    public static final String S3_UPLOAD_BUCKET_KEY = "S3_UPLOAD_BUCKET";
+
+    //    public final transient String clientRegion;
+    public final transient String bucketName;
+    private final transient String allowedOrigin;
+    private final transient AmazonS3 s3Client;
+
+    public CreateUploadHandler() {
+        this(new Environment(), createAmazonS3Client());
+    }
+
+
+    /**
+     * Construct for lambda eventhandler to create an upload request for S3.
+     */
+    public CreateUploadHandler(Environment environment, AmazonS3 s3Client) {
+        this.allowedOrigin = environment.get(ALLOWED_ORIGIN_KEY).orElseThrow(IllegalStateException::new);
+        this.bucketName = environment.get(S3_UPLOAD_BUCKET_KEY).orElseThrow(IllegalStateException::new);
+        this.s3Client = s3Client;
+    }
+
+    public static AmazonS3 createAmazonS3Client() {
+        return AmazonS3ClientBuilder.standard()
+                .build();
+    }
+
+    @Override
+    public GatewayResponse handleRequest(Map<String, Object> input, Context context) {
+
+        final GatewayResponse response = new GatewayResponse();
+        response.setHeaders(headers());
+
+
+        CreateUploadRequestBody requestBody = null;
+        try {
+            requestBody = checkParameters(input);
+        } catch (JsonSyntaxException | ParameterMissingException e) {
+            response.setErrorBody(e.getMessage());
+            response.setStatusCode(SC_BAD_REQUEST);
+            System.out.println(response);
+            return response;
+        } catch (Exception e) {
+            response.setErrorBody(e.getMessage());
+            response.setStatusCode(SC_INTERNAL_SERVER_ERROR);
+            System.out.println(response);
+            return response;
+        }
+
+        ObjectMetadata objectMetadata = getObjectMetadata(requestBody);
+
+        String keyName = UUID.randomUUID().toString();
+        InitiateMultipartUploadRequest initRequest =
+                new InitiateMultipartUploadRequest(bucketName, keyName, objectMetadata);
+
+        try {
+            InitiateMultipartUploadResult initResponse = s3Client.initiateMultipartUpload(initRequest);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

This code manually manages uploads or downloads. The SDK can automatically handle these operations. Consider using the SDK to automatically handle this instead. SDK class: TransferManager. See [Using TransferManager for Amazon S3 Operations](https://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/examples-s3-transfermanager.html).